core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -369,19 +400,29 @@ public void halfClose() {
     Preconditions.checkState(stream != null, "Not started");
     Preconditions.checkState(!cancelCalled, "call was cancelled");
     Preconditions.checkState(!halfCloseCalled, "call already half-closed");
+    if (!streamStarted) {
+      cancel("No request message has been sent", null);
+      return;
+    }
     halfCloseCalled = true;
     stream.halfClose();
   }
 
   @Override
   public void sendMessage(ReqT message) {
-    Preconditions.checkState(stream != null, "Not started");
     Preconditions.checkState(!cancelCalled, "call was cancelled");
     Preconditions.checkState(!halfCloseCalled, "call was half-closed");
     try {
       // TODO(notcarl): Find out if messageIs needs to be closed.
       InputStream messageIs = method.streamRequest(message);
-      stream.writeMessage(messageIs);
+      if (!useGetMethod) {
+        stream.writeMessage(messageIs);
+      } else {
+        // Encode the request message and put it in the request headers.
+        byte[] payload = ByteStreams.toByteArray(messageIs);
+        headers.put(GrpcUtil.GRPC_PAYLOAD_BIN_KEY, payload);
+        startStream();
```

Comment:
Can you delay start until half close? half close seems a much more natural place to start the stream (delayed, like this) than on the message.